### ISSUE TITLE
realpath called with unchecked arguments nullifies Docker image building support on NixOS

### ISSUE THREAD

**Issue by spy-tech** [NixOS/nixpkgs#21100]
The following applies to both 16.09 and master:

realpath can be called with a missing argument like can be observed in the following output: 

```
Unpacking base image...
realpath: missing operand
Try 'realpath --help' for more information.
realpath: missing operand
Try 'realpath --help' for more information.
realpath: missing operand
Try 'realpath --help' for more information.
realpath: missing operand
Try 'realpath --help' for more information.
```

The location of the call is here. 

```
# build-support/docker/default.nix
421:          for f in $(tar -tf $1 | xargs realpath -ms --relative-to=.); do
```

I'd say the basic problem is that the above code is written in an *extremely* fragile way. The reference to $1 is incorrect, looping in this particular way is also wrong in general, assuming the set of files that is being generated is probably also wrong in general. There is just so much wrong with it, that it's hard to list all of it. 

This breaks docker support on NixOS completely and as such is a blocker.

**Comment by joachifm** [comment]
cc @adnelson (via git blame)

**Comment by adnelson** [comment]
I have my hands all over the git blame due to a readability and documentation refactor I did a while back but unfortunately the actual logic of the functions was mostly a blind copypaste on my part :( I could take a look at what's going on but my insight wouldn't be any better than anyone else.

**Comment by joachifm** [comment]
@adnelson ah, I'm sorry, I didn't look too closely. @tboettch may know more

**Comment by tboettch** [comment]
I'm not finding what you're referring to in [16.09](https://github.com/NixOS/nixpkgs/blob/16.09/pkgs/build-support/docker/default.nix). I do see it in master.

My original involvement here was fixing an issue where layers with shared content were not being de-duplicated, leading to larger than necessary layers (see #18896, the commit message has details). I'm not a Bash expert, so if you have some constructive feedback or a patch to offer, I'd be happy to review it. That said, I'm going to be sparsely available until early January, so please don't wait on my account.

I will say that it was working fine for me the last time I used it. Maybe you can elaborate on what sorts of inputs cause this kind of breakage? I've only used base images generated by Nix itself, so I imagine the problems arise when importing external images? If there are test cases that could be added here, they would help avoid regressions in the future.

**Comment by Mic92** [comment]
there way too much shell in this module given the complexity of the task. If I would use it, I would rewrite in a different language, like I did for https://github.com/Mic92/nix2aci

**Comment by Profpatsch** [comment]
(triage) Is this still a problem with the current `dockerTools`?

**Comment by Mic92** [comment]
There have been a lot of changes in this code since than. Please open a new issue if this is still a problem.
